Data/Primitive/MVar.hs

```diff
@@ -0,0 +1,133 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+-- |
+-- Module      : Data.Primitive.ByteArray
+-- Copyright   : (c) Roman Leshchinskiy 2009-2012
+-- License     : BSD-style
+--
+-- Maintainer  : Roman Leshchinskiy <rl@cse.unsw.edu.au>
+-- Portability : non-portable
+--
+-- Primitive operations on @MVar@. This module provides a similar interface
+-- to "Control.Concurrent.MVar". However, the functions are generalized to
+-- work in any 'PrimMonad' instead of only working in 'IO'. Note that all
+-- of the functions here are completely deterministic. It is only the
+-- introduction of @forkIO@ that brings nondeterminism into the picture.
+-- Users who are interesting in using these functions in 'ST' will likely
+-- want to roll their own 'ST' variant of @forkIO@. Since such a function
+-- eschews that determinism expected by users of 'ST' (and users of
+-- 'primitive'), it is left to the user to implement it. 
+module Data.Primitive.MVar
+  ( MVar(..)
+  , newMVar
+  , isEmptyMVar
+  , newEmptyMVar
+  , putMVar
+#if __GLASGOW_HASKELL__ >= 708
+  , readMVar
+#endif
+  , takeMVar
+  , tryPutMVar
+#if __GLASGOW_HASKELL__ >= 708
+  , tryReadMVar
+#endif
+  , tryTakeMVar
+  ) where
+
+import Control.Monad.Primitive
+import Data.Primitive.Internal.Compat (isTrue#)
+import GHC.Exts (MVar#,newMVar#,takeMVar#,sameMVar#,putMVar#,tryTakeMVar#,
+  isEmptyMVar#,tryPutMVar#,(/=#))
+
+#if __GLASGOW_HASKELL__ >= 708
+import GHC.Exts (readMVar#,tryReadMVar#)
+#endif
+
+data MVar s a = MVar (MVar# s a)
+
+instance Eq (MVar s a) where
+  (MVar mvar1#) == (MVar mvar2#) = isTrue# (sameMVar# mvar1# mvar2#)
+
+-- | Create an 'MVar' which is initially empty.
+newEmptyMVar :: PrimMonad m => m (MVar (PrimState m) a)
+newEmptyMVar = primitive $ \ s# ->
+  case newMVar# s# of
+    (# s2#, svar# #) -> (# s2#, MVar svar# #)
+
+
+-- |Create an 'MVar' which contains the supplied value.
+newMVar :: PrimMonad m => a -> m (MVar (PrimState m) a)
+newMVar value =
+  newEmptyMVar >>= \ mvar ->
+  putMVar mvar value >>
+  return mvar
+
+-- | Return the contents of the 'MVar'.  If the 'MVar' is currently
+-- empty, 'takeMVar' will wait until it is full.  After a 'takeMVar',
+-- the 'MVar' is left empty.
+takeMVar :: PrimMonad m => MVar (PrimState m) a -> m a
+takeMVar (MVar mvar#) = primitive $ \ s# -> takeMVar# mvar# s#
+
+#if __GLASGOW_HASKELL__ >= 708
+-- | Atomically read the contents of an 'MVar'.  If the 'MVar' is
+-- currently empty, 'readMVar' will wait until it is full.
+-- 'readMVar' is guaranteed to receive the next 'putMVar'.
+--
+-- This function is only available when compiling with GHC 7.8
+-- or newer.
+readMVar :: PrimMonad m => MVar (PrimState m) a -> m a
+readMVar (MVar mvar#) = primitive $ \ s# -> readMVar# mvar# s#
+#endif
```

Comment:
prior to 7.8 read mvar was encoded in terms of take then put, granted prior to 7.8 it would block other reads, which wasn't ideal...

whats the oldest ghc we currently support?

Reply:
We currently support back to 7.4.

Reply:
Dang. I just realized that it's actually just not possible to shim this in a way that has the same behavior. I'm going to add a note on the main thread about this.

Reply:
Another option is to use a different representation. When compiling with old GHC, we can do this:

```haskell
data Use = Reading | Taking

data MVar s a = MVar
  { var :: !(STRef s (Maybe a))
  , claimed :: MVar# s Use }
```

Horrible, right? Now `takeMVar` will stake a claim by *putting* `Taking` into the `MVar#`, and then set the `STRef` to `Nothing`. `putMVar` will *take* the `MVar#`. If it's `Taking`, it will then set the `STRef`. If it's `Reading`, it will instead put it back and `yield`. `readMVar` will put `Reading` into the `MVar#`, read the `STRef`, and then take the `MVar#` again.

I don't really think this is a good idea, but I figured I'd mention it.